File: androidApp/src/main/kotlin/com/humblesolutions/aromex/data/BackendCompanySettingsRepository.kt
package com.humblesolutions.aromex.data

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.SetOptions
import com.google.firebase.ktx.Firebase
import com.google.firebase.firestore.ktx.firestore
import android.content.Context
import com.humblesolutions.aromex.model.CompanyDetails
import com.humblesolutions.aromex.model.CompanySettingsChange
import com.humblesolutions.aromex.model.FirebaseClientConfig
import com.humblesolutions.aromex.model.TaxConfig
import com.humblesolutions.aromex.repository.CompanyProfile
import com.humblesolutions.aromex.repository.CompanySettingsRepository
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow

/**
 * Firestore-backed [CompanySettingsRepository] for Android — `companySettings/profile`.
 *
 * Added so a phone's till follows a tax-rate change (M10). Without it, `UserSession.tax` is captured
 * at sign-in and never refreshed: an admin changing GST at noon would leave every open phone
 * charging the old rate for the rest of the day, silently — the same hole that was closed on
 * Desktop.
 *
 * Reads only. The Settings *screen* is Desktop-only for now, so the write paths throw rather than
 * pretend: a phone that appeared to save a tax rate but didn't would be worse than one that can't.
 */
class BackendCompanySettingsRepository(
    private val context: Context,
    private val config: FirebaseClientConfig,
) : CompanySettingsRepository {

    private val app get() = FirebaseAppFactory.get(context, config)
    private val db get() = Firebase.firestore(app)

    override fun observeProfile(): Flow<CompanyProfile> = callbackFlow {
        val registration = db.collection(SETTINGS).document(PROFILE)
            .addSnapshotListener { snapshot, error ->
                when {
                    error != null -> close(error)
                    snapshot != null && snapshot.exists() -> trySend(snapshot.toProfile())
                    // An absent profile means provisioning hasn't finished. Emitting a default
                    // would quietly say "no tax"; staying silent lets the caller keep what it had.
                    else -> Unit
                }
            }
        awaitClose { registration.remove() }
    }

    override fun observeChanges(limit: Int): Flow<List<CompanySettingsChange>> = callbackFlow {
        val registration = db.collection(CHANGES)
            .orderBy("changedAt", Query.Direction.DESCENDING)
            .limit(limit.toLong())
            .addSnapshotListener { snapshot, error ->
                when {
                    error != null -> close(error)
                    snapshot != null -> trySend(snapshot.documents.mapNotNull { it.toChange() })
                    else -> trySend(emptyList())
                }
            }
        awaitClose { registration.remove() }
    }

    override suspend fun updateTax(config: TaxConfig, changedBy: String, changedByName: String?) =
        throw UnsupportedOperationException("Tax settings are edited on Desktop")

    override suspend fun updateDetails(details: CompanyDetails, changedBy: String, changedByName: String?) =
        throw UnsupportedOperationException("Company details are edited on Desktop")

    // ── mapping (mirrors the Desktop implementation) ─────────────────────────

    private fun DocumentSnapshot.toProfile() = CompanyProfile(
        hlCompanyId = getString("hlCompanyId").orEmpty(),
        currency = getString("currency").orEmpty(),
        companyName = getString("companyName").orEmpty(),
        tax = parseTax(get("tax")),
        timezone = getString("timezone")?.takeIf { it.isNotBlank() } ?: "UTC",
        legalName = getString("legalName"),
        taxNumber = getString("taxNumber"),
        logoUrl = getString("logoUrl"),
        businessAddress = getString("businessAddress"),
        contactEmail = getString("contactEmail"),
        contactPhone = getString("contactPhone"),
    )

    /** Rates may be stored as a number (provisioning) or a string (the Settings screen) — accept both. */
    private fun parseTax(raw: Any?): TaxConfig {
        val tax = raw as? Map<*, *> ?: return TaxConfig()
        fun rate(v: Any?): String = when (v) {
            is String -> v.trim().ifEmpty { "0" }
            is Number -> v.toString()
            else -> "0"
        }
        return TaxConfig(
            gstEnabled = tax["gstEnabled"] as? Boolean ?: false,
            gstRate = rate(tax["gstRate"]),
            pstEnabled = tax["pstEnabled"] as? Boolean ?: false,
            pstRate = rate(tax["pstRate"]),
            isHST = tax["isHST"] as? Boolean ?: false,
        )
    }

    /** Null when the row is unreadable — a malformed entry is skipped, never guessed at. */
    private fun DocumentSnapshot.toChange(): CompanySettingsChange? {
        val field = getString("field") ?: return null
        val new = getString("newValue") ?: return null
        return CompanySettingsChange(
            changeId = id,
            field = field,
            oldValue = getString("oldValue"),
            newValue = new,
            changedBy = getString("changedBy").orEmpty(),
            changedByName = getString("changedByName"),
            changedAt = getDate("changedAt")?.time ?: 0L,
        )
    }

    private companion object {
        const val SETTINGS = "companySettings"
        const val PROFILE = "profile"
        const val CHANGES = "companySettingsChanges"
    }
}

File: iosApp/iosApp/repository/BackendCompanySettingsRepository.swift
import Foundation
import FirebaseFirestore
import SharedLogic

/// Firestore-backed implementation of the shared `CompanySettingsRepository` — `companySettings/profile`.
///
/// ── Why this exists ──────────────────────────────────────────────────────────
/// So a phone's till follows a tax-rate change. `UserSession.tax` is captured at sign-in and never
/// refreshed, so without a live profile an admin changing GST at noon would leave every open phone
/// charging the old rate for the rest of the day, silently — the hole closed on Desktop (M10) and
/// Android, and now here.
///
/// ── Flow-interop (SKIE) ──────────────────────────────────────────────────────
/// `observeProfile()` returns a Kotlin `Flow<CompanyProfile>`, built with the shared
/// `companyProfileCallbackFlow(subscribe:)` adapter feeding a Firestore snapshot listener — the
/// same shape as `BackendCommissionRuleRepository` and `BackendEntityRepository`. SKIE
/// `__`-prefixes the suspend members.
///
/// Reads only. The Settings *screen* is Desktop-only, so the write paths throw rather than
/// pretend: a phone that appeared to save a tax rate but didn't would be worse than one that
/// plainly can't.
final class BackendCompanySettingsRepository: NSObject, CompanySettingsRepository {

    private let db: Firestore

    init(config: FirebaseClientConfig) {
        self.db = FirebaseAppFactory.firestore(for: config)
        super.init()
    }

    // MARK: - Live read (produces the shared Flow the use case consumes)

    func observeProfile() -> SkieSwiftFlow<CompanyProfile> {
        return companyProfileCallbackFlow { onEach, onError in
            let registration = self.db.collection("companySettings").document("profile")
                .addSnapshotListener { snapshot, error in
                    if let error = error {
                        _ = onError(error.localizedDescription)
                        return
                    }
                    // An absent profile means provisioning hasn't finished. Emitting a default
                    // would quietly say "no tax"; staying silent lets the caller keep what it had.
                    guard let data = snapshot?.data(), snapshot?.exists == true else { return }
                    _ = onEach(Self.mapProfile(data))
                }
            return FirestoreCompanyProfileObservation(registration)
        }
    }

    /// The change log is part of the Desktop-only Settings screen; an empty history is the honest
    /// answer here rather than a pretence that one was fetched.
    func observeChanges(limit: Int32) -> SkieSwiftFlow<[CompanySettingsChange]> {
        return emptySettingsChangesFlow()
    }

    // MARK: - Writes (Desktop-only)

    func __updateTax(config: TaxConfig, changedBy: String, changedByName: String?) async throws {
        throw NSError(
            domain: "Aromex", code: 1,
            userInfo: [NSLocalizedDescriptionKey: "Tax settings are edited on Desktop"])
    }

    func __updateDetails(details: CompanyDetails, changedBy: String, changedByName: String?) async throws {
        throw NSError(
            domain: "Aromex", code: 1,
            userInfo: [NSLocalizedDescriptionKey: "Company details are edited on Desktop"])
    }

    // MARK: - Mapping (mirrors the Desktop and Android implementations)

    private static func mapProfile(_ d: [String: Any]) -> CompanyProfile {
        CompanyProfile(
            hlCompanyId: d["hlCompanyId"] as? String ?? "",
            currency: d["currency"] as? String ?? "",
            companyName: d["companyName"] as? String ?? "",
            tax: mapTax(d["tax"]),
            timezone: (d["timezone"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "UTC",
            legalName: d["legalName"] as? String,
            taxNumber: d["taxNumber"] as? String,
            logoUrl: d["logoUrl"] as? String,
            businessAddress: d["businessAddress"] as? String,
            contactEmail: d["contactEmail"] as? String,
            contactPhone: d["contactPhone"] as? String)
    }

    /// Rates may be stored as a number (provisioning) or a string (the Settings screen) — accept
    /// both, and never let a rate become a Double on the way through.
    private static func mapTax(_ raw: Any?) -> TaxConfig {
        guard let t = raw as? [String: Any] else {
            return TaxConfig(gstEnabled: false, gstRate: "0", pstEnabled: false, pstRate: "0", isHST: false)
        }
        func rate(_ v: Any?) -> String {
            if let s = v as? String {
                let trimmed = s.trimmingCharacters(in: .whitespaces)
                return trimmed.isEmpty ? "0" : trimmed
            }
            if let n = v as? NSNumber { return n.stringValue }
            return "0"
        }
        return TaxConfig(
            gstEnabled: t["gstEnabled"] as? Bool ?? false,
            gstRate: rate(t["gstRate"]),
            pstEnabled: t["pstEnabled"] as? Bool ?? false,
            pstRate: rate(t["pstRate"]),
            isHST: t["isHST"] as? Bool ?? false)
    }
}

/// Detaches the Firestore listener when the shared Flow is cancelled.
private final class FirestoreCompanyProfileObservation: NSObject, CompanyProfileObservation {
    private let registration: ListenerRegistration
    init(_ registration: ListenerRegistration) { self.registration = registration }
    func cancel() { registration.remove() }
}
